odoo/models.py

```diff
@@ -3413,6 +3426,22 @@ def unlink(self):
         self.check_access_rights('unlink')
         self._check_concurrency()
 
+        from odoo.addons.base.models.ir_model import MODULE_UNINSTALL_FLAG
+        # go through _ondelete_methods in reverse order:
+        # let module A be a module that implements an _ondelete_method X for model M
+        # let module B be a module that extends A's model M and introduces _ondelete_method Y
+        # let both conditions for X and Y be True and trigger the respective exceptions
+        # if module B introduces tests that check for Y's exception to be raised, and the
+        # record on which the test is being performed happens to also trigger X's exception (or
+        # any other number of exceptions implemented elsewhere that the developer of B is not aware
+        # of), then the test will fail because X's exception will trigger first and the test
+        # will never reach Y's exception. By reversing the order we guarantee that the exception
+        # raised (if multiple) is the one implemented by the last extending module.
+        for func in reversed(self._ondelete_methods):
```

Comment:
:warning: this explanation is not correct!  The list `self._ondelete_methods` is retrieved with `inspect.getmembers`, which returns all the members of an object in a list of `(name, value)` pairs sorted by `name`.  Using `reversed` here might work for one use case, and not for another.  If a test fails because of this, then it means that the test is not correct...